### ISSUE TITLE
Improved narrowing by literal expression

### ISSUE THREAD

**Issue by ethframe** [microsoft/pyright#2817]
**Is your feature request related to a problem? Please describe.**

I have the following code that is accepted by mypy:

```python
from typing import Literal, Union, final


@final
class A:
    def __init__(self, x: bool, y: int):
        self.x = x
        self.y = y


@final
class B:
    def __init__(self) -> None:
        self.x: Literal[False] = False


X = Union[A, B]


def fn(x: X) -> int:
    if x.x is True:
        reveal_type(x)
        return x.y
    else:
        reveal_type(x)
        return 0
```

```
mypy --strict fn.py
fn.py:22: note: Revealed type is "fn.A"
fn.py:25: note: Revealed type is "Union[fn.A, fn.B]"
```

pyright fails to narrow the type of `x` in the first arm of `if`:

```
$ pyright fn.py
...
...\fn.py
  ...\fn.py:22:21 - info: Type of "x" is "A | B"
  ...\fn.py:23:18 - error: Cannot access member "y" for type "B"
    Member "y" is unknown (reportGeneralTypeIssues)
  ...\fn.py:23:16 - error: Type of "y" is unknown (reportUnknownMemberType)
  ...\fn.py:23:16 - error: Type of "y" is partially unknown
    Type of "y" is "int | Unknown" (reportUnknownMemberType)
  ...\fn.py:23:16 - error: Return type, "int | Unknown", is partially unknown (reportUnknownVariableType)
  ...\fn.py:25:21 - info: Type of "x" is "A | B"
4 errors, 0 warnings, 2 infos
```

**Describe the solution you'd like**

To be consistent with mypy, pyright should treat type guards of form `x.y is L` as identical to `x.y == L` if `L` is a singleton, and it should narrow type even when some literal value is used for two or more types.

**Additional context**

```
$ pyright --version
pyright 1.1.204
$ mypy --version
mypy 0.930
```

**Comment by erictraut** [comment]
Thanks for the enhancement request. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.205, which I just published. It will also be included in the next release of pylance.
